lib/new_relic/agent/log_event_aggregator.rb

```diff
@@ -228,6 +238,16 @@ def truncate_message(message)
         return message if message.bytesize <= MAX_BYTES
         message.byteslice(0...MAX_BYTES)
       end
+
+      # don't forward logs from a logger writing to a standard stream
+      # (stdout/stderr) unless it happens to be the only logger in use
+      def mark_standard_streams_for_skipping
+        return unless @loggers.keys.size > 1
+
+        @loggers.each do |logger, dev|
+          logger.mark_skip_instrumenting if dev.class == IO
+        end
+      end
```

Comment:
I think I might not be fully understanding this solution. I'm concerned about how things will be have if, for example we have the following sources/output locations: 

| source | output location |
| ------ | ---------------- |
| rails | production.log |
| serveza | serveza.log |
| mail | STDOUT |

Would the mail logs be marked with `@skip_instrumenting`? This gems outputs only to STDOUT.

Maybe it's the construction of the hash that I can't quite see. I think I'm hoping for a data type more like this: 

`{:rails => ['production.log'], :serveza => ['serveza.log'], :mail => ['STDOUT']}` so that as long as each source only has one item in the array, it will be kept. If the array is longer than one, the logger is marked `@skip_instrumenting`. Is that how this works?

Reply:
The hash key is a reference to an instance of `::Logger`. The hash value is a reference to that logger's underlying output handle (the `::Logger` instance has a `@logdev` instance var, and that `@logdev` object has a `@log` instance var, and it's the `@dev` object's reference we're storing in the hash.

So it's just a 2 dimensional hash with references for both the keys and values (and no arrays anywhere).

The logic is...

* When an unregistered `::Logger` instance writes for the first time, register it.
* When registering, check the registration hash. If this logger is the only logger, short circuit.
* When registering, if multiple loggers exist in the hash, then mark all ones that write to `IO` for skipping.

NOTE that `klass == IO` will match $stdout, $stderr, STDOUT.dup, etc., but not an instance of `File` (which itself would pass an `#is_a?(IO)` check).

So we end up with a hash that tells us of all the `::Logger` objects we've seen (skipped or otherwise) and the device they're writing to.

I'm using this approach to take advantage of the fact that we effectively have a singleton class to track the hash with and we can avoid setting class vars on the `::Logger` class itself.

Reply:
Ok, thanks for clarifying. I think I want to run a quick test with the Mail library on a playground app to confirm those logs make it through this setup before approving. If that works, then I'll feel more at ease.